xarray/tests/test_dask.py

```diff
@@ -917,3 +918,176 @@ def test_dask_layers_and_dependencies():
+    map_ds.coords["cxy"] = map_ds.cxy.chunk({"y": 10})
+    map_ds.attrs["test"] = "test"
+    map_ds["xx"] = (map_ds["a"] * map_ds.y).chunk({"y": 20})
+
+    return map_ds
+
+
+# work around mypy error
+# xarray/tests/test_dask.py:888: error: Dict entry 0 has incompatible type "str": "int"; expected "Hashable": "Union[None, Number, Tuple[Number, ...]]"
+map_da = make_da()
+map_ds = make_ds()
+
+
+def test_unify_chunks():
+    with raises_regex(ValueError, "inconsistent chunks"):
+        map_ds.chunks
+
+    expected = {"x": (4, 4, 2), "y": (5, 5, 5, 5)}
+    actual = map_ds.unify_chunks().chunks
+    assert expected == actual
+
+
+# TODO: DataArray.chunks is not a dict but Dataset.chunks is!
+def assert_chunks_equal(a, b):
+
+    # unchunked dimensions in input to map_blocks become 1 chunk.
+    # do this manually before comparing
+    def at_least_one_chunk(obj):
+        new_chunks = {}
+        for dim in obj.dims:
+            if dim not in obj.chunks:
+                new_chunks[dim] = len(obj[dim])
+            else:
+                # must preserve old chunks too
+                new_chunks[dim] = obj.chunks[dim][0]
+        return obj.chunk(new_chunks)
+
+    if isinstance(a, DataArray):
+        a = a._to_temp_dataset()
+
+    if isinstance(b, DataArray):
+        b = b._to_temp_dataset()
+
+    a = at_least_one_chunk(a.unify_chunks())
+    b = at_least_one_chunk(b.unify_chunks())
+
+    assert a.chunks == b.chunks
+
+
+def test_map_blocks_error():
+    def bad_func(darray):
+        return (darray * darray.x + 5 * darray.y)[:1, :1]
+
+    with raises_regex(ValueError, "Length of the.* has changed."):
+        xr.map_blocks(bad_func, map_da).compute()
+
+    def returns_numpy(darray):
+        return (darray * darray.x + 5 * darray.y).values
+
+    with raises_regex(ValueError, "Function must return an xarray DataArray"):
+        xr.map_blocks(returns_numpy, map_da)
+
+    with raises_regex(ValueError, "args must be"):
+        xr.map_blocks(operator.add, map_da, args=10)
+
+    with raises_regex(ValueError, "kwargs must be"):
+        xr.map_blocks(operator.add, map_da, args=[10], kwargs=[20])
+
+
+@pytest.mark.parametrize("obj", [map_da, map_ds])
+def test_map_blocks(obj):
+    def func(obj):
+        result = obj + obj.x + 5 * obj.y
+        return result
+
+    with raise_if_dask_computes():
+        actual = xr.map_blocks(func, obj)
+    expected = func(obj).unify_chunks()
+    assert_chunks_equal(expected, actual)
+    # why is compute needed?
+    xr.testing.assert_equal(expected.compute(), actual.compute())
+
+
+@pytest.mark.parametrize("obj", [map_da, map_ds])
+def test_map_blocks_convert_args_to_list(obj):
+    expected = obj.unify_chunks() + 10
+    with raise_if_dask_computes():
+        actual = xr.map_blocks(operator.add, obj, [10])
+    assert_chunks_equal(expected, actual)
+    # why is compute needed?
+    xr.testing.assert_equal(expected.compute(), actual.compute())
```

Comment:
indeed it should not be needed! please investigate

Reply:
`assert_equal(actual, expected)` passes but `assert_equal(expected, actual)` fails! See #3350 for some weird behaviour.